sys/shell_lock/shell_lock.c

```diff
@@ -0,0 +1,201 @@
+ * After 3 (default) failed attempts, the input is blocked for a few seconds to
+ * slow down brute force attacks.
+ * Does not make use of any cryptographic features yet.
+ *
+ * This module also provides a pseudomodule for automated locking after a given
+ * interval. Add "USEMODULE += shell_lock_auto_locking" to your Makefile to
+ * enable this feature.
+ *
+ * @author      Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include "xtimer.h"
+
+#include "shell_lock.h"
+
+static bool _shell_is_locked = true;
+
+#ifdef MODULE_SHELL_LOCK_AUTO_LOCKING
+static xtimer_t _shell_auto_lock_xtimer;
+#endif
+
+/* defined in shell.c */
+extern void flush_if_needed(void);
+
+/* defined in shell.c */
+extern void print_prompt(void);
+
+/* defined in shell.c */
+extern int readline(char *buf, size_t size);
+
+int _lock_handler(int argc, char **argv)
+{
+    (void) argc;
+    (void) argv;
+
+    _shell_is_locked = true;
+
+    return 0;
+}
+
+/* This should be independent from the module "shell_commands", as the user may
+ * don't want any other command than the mandatory "lock"-command. */
+const shell_command_t _shell_lock_command_list[] = {
+        {"lock", "Lock the shell", _lock_handler},
+        {NULL, NULL, NULL}
+};
+
+/* Implementation of strcmp that does not return after the first difference
+ * which could give away information about the first n correct characters of
+ * the password. The length of the loop is only dependent on the input string.
+ * Don't optimize this function by a compiler. */
+static bool __attribute__((optimize("O0"))) _safe_strcmp(char* input, char* pwd)
+{
+    bool the_same = true;
+
+    int input_len = strlen(input);
+    int pwd_len = strlen(pwd);
+
+    int input_index = 0;
+    int pwd_index = 0;
+
+    do {
+        if (input[input_index] != pwd[pwd_index]) {
+            the_same &= false;
+        }
+        else {
+            the_same &= true;
+        }
+
+        /* keep indices at last index of respective string */
+        if (input_index < input_len) {
+            input_index++;
+        }
+
+        if (pwd_index < pwd_len) {
+            pwd_index++;
+        }
+
+    } while (input[input_index] != '\0');
+
+    if (input_len != pwd_len) {
+        /* substring of the password doesn't count */
+        return false;
+    }
+
+    return the_same;
+}
+
+static bool _login(char *line_buf, size_t buf_size)
+{
+    int success = false;
+
+    puts("Password:");
+
+    print_prompt();
+
+    if (readline(line_buf, buf_size) > 0) {
+
+        if (_safe_strcmp(line_buf, SHELL_LOCK_PASSWORD)) {
+            success = true;
+        }
+    }
+
+    return success;
+}
+
+/**
+ * Repeatedly prompt for the password.
+ *
+ * This function won't return until the correct password has been entered.
+ */
+void _login_barrier(char *line_buf, size_t buf_size)
```

Comment:
privater functions should be 
```suggestion
static void _login_barrier(char *line_buf, size_t buf_size)
```

Reply:
check